Warn and skip on glacier objects that may fail for s3 commands

kyleknap opened the pull request:
This pull request does the following:
* Skips operations that involve transferring a glacier object. This includes downloading, copying, and moving an object when the source is a glacier object. If the glacier object is skipped, a warning is displayed causing the return code to be ``2``. The most important part of this functionality is that we will not even try performing an operation, which could slow down a command especially if you have a lot of or large glacier objects.
* Add ``--ignore-glacier-warnings`` argument that allows you to hide the glacier warnings and not cause the return code to be 2 if a glacier object is encountered. Note that glacier objects are still always skipped if the command involves downloading, copying, or moving an object when the source is a glacier object.

All of the integration tests pass.

Fixes https://github.com/aws/aws-cli/issues/748

cc @jamesls @mtdowling @rayluo @JordonPhillips

jamesls commented on awscli/customizations/s3/filegenerator.py:
```diff
@@ -324,4 +326,4 @@ def _list_single_object(self, s3_path):
         file_size = int(response['ContentLength'])
         last_update = parse(response['LastModified'])
         last_update = last_update.astimezone(tzlocal())
-        return s3_path, file_size, last_update
```
This seems like a weird API.  We're returning parts of the response (such as LastModified) as well as just the entire response whole sale.  This can probably be cleaned up.

kyleknap replied:
I kept it like this to keep parity with the output for local file listings. There is no response associated with a local listing. I will probably break up the logic such that the listing of files and listing of s3 objects do not have the same interface.

jamesls commented on awscli/customizations/s3/utils.py:
```diff
@@ -353,6 +353,44 @@ def relative_path(filename, start=os.path.curdir):
         return os.path.abspath(filename)
 
 
+def is_glacier_object(response_data):
```
Is there a reason this needs to be extracted out?  Wouldn't it make more sense in fileinfo?

kyleknap replied:
I extracted it because I saw it as being potential useful if you were to write a glacier-specific sync strategy. I am fine with moving it to FileInfo though since we are not using it anywhere else.

mtdowling replied:
Yeah, +1 on moving these to FileInfo.

mtdowling commented on awscli/customizations/s3/filegenerator.py:
```diff
@@ -157,7 +159,7 @@ def list_files(self, path, dir_op):
             if not dir_op:
                 size, last_update = get_file_stat(path)
-                yield path, size, last_update
+                yield path, size, last_update, None
```
Returning tuples with four elements, some of which are `None` seems like a code smell. Perhaps this method (and other similar methods) should return a dict or a specific object instead of a long tuple?

kyleknap replied:
Yeah I was trying to touch the least amount of code paths as possible, but I agree it is getting too messy. I am going to just refactor the ``FileGenerator`` class to simplify some of the interfaces. Should not be hard.

kyleknap commented:
Alright, I think I incorporated the feedback. Refactored the ``FileGenerator`` a bit. Mainly it allowed the ``list_files`` and ``list_objects`` to yield data that may differ as opposed to forcing the two methods to return the same size tuples that may or may not be completely filled. Ready to get looked at again.

kyleknap commented:
Alright I cleaned up the code a bit more. Should be good to look at.

jamesls commented:
:shipit:

mtdowling commented on awscli/customizations/s3/subcommands.py:
```diff
@@ -290,12 +290,23 @@
+    'help_text': (
+        'Turns off glacier warnings. Warnings about operations that cannot '
+        'be performed because it involves copying, downloading, or moving '
+        'a glacier object will no longer be printed to standard error and '
```
Changing this to "*and will no longer* cause the return code of the command to be ``2``" might make it a bit more clear that using this option will prevent an error exit rather than cause one.

kyleknap replied:
Updated

mtdowling commented:
:ship:
